// twitch/lib/API/Channel/ChannelAPI.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Decorators_1 = require("../../Toolkit/Decorators");
var BaseAPI_1 = require("../BaseAPI");
var _1 = require("./");
var UserTools_1 = require("../../Toolkit/UserTools");
var ChannelSubscription_1 = require("./ChannelSubscription");
var errors_1 = require("request-promise-native/errors");
var NotSubscribed_1 = require("../NotSubscribed");
var NoSubscriptionProgram_1 = require("../NoSubscriptionProgram");
var PrivilegedChannel_1 = require("./PrivilegedChannel");
var _2 = require("../User/");
var ChannelFollow_1 = require("./ChannelFollow");
var ChannelAPI = /** @class */ (function (_super) {
    __extends(ChannelAPI, _super);
    function ChannelAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChannelAPI.prototype.getMyChannel = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = PrivilegedChannel_1.default.bind;
                        return [4 /*yield*/, this._client.apiCall({ url: 'channel', scope: 'channel_read' })];
                    case 1: return [2 /*return*/, new (_a.apply(PrivilegedChannel_1.default, [void 0, _b.sent(), this._client]))()];
                }
            });
        });
    };
    ChannelAPI.prototype.getChannel = function (user) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = _1.default.bind;
                        return [4 /*yield*/, this._client.apiCall({ url: "channels/" + UserTools_1.default.getUserId(user) })];
                    case 1: return [2 /*return*/, new (_a.apply(_1.default, [void 0, _b.sent(), this._client]))()];
                }
            });
        });
    };
    ChannelAPI.prototype.updateChannel = function (channel, data) {
        return __awaiter(this, void 0, void 0, function () {
            var channelId;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        channelId = UserTools_1.default.getUserId(channel);
                        return [4 /*yield*/, this._client.apiCall({
                                url: "channels/" + channelId,
                                method: 'PUT',
                                jsonBody: { channel: data },
                                scope: 'channel_editor'
                            })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ChannelAPI.prototype.getChannelEditors = function (channel) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var channelId, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        channelId = UserTools_1.default.getUserId(channel);
                        return [4 /*yield*/, this._client.apiCall({
                                url: "channels/" + channelId + "/editors",
                                scope: 'channel_read'
                            })];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.users.map(function (userData) { return new _2.default(userData, _this._client); })];
                }
            });
        });
    };
    ChannelAPI.prototype.getChannelFollowers = function (channel, page, limit, orderDirection) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var channelId, query, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        channelId = UserTools_1.default.getUserId(channel);
                        query = {};
                        if (page) {
                            query.offset = ((page - 1) * (limit || 25)).toString();
                        }
                        if (limit) {
                            query.limit = limit.toString();
                        }
                        if (orderDirection) {
                            query.direction = orderDirection;
                        }
                        return [4 /*yield*/, this._client.apiCall({
                                url: "channels/" + channelId + "/follows",
                                query: query
                            })];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.follows.map(function (follow) { return new ChannelFollow_1.default(follow, _this._client); })];
                }
            });
        });
    };
    ChannelAPI.prototype.getChannelSubscriptions = function (channel, page, limit, orderDirection) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            var channelId, query, data, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        channelId = UserTools_1.default.getUserId(channel);
                        query = {};
                        if (page) {
                            query.offset = ((page - 1) * (limit || 25)).toString();
                        }
                        if (limit) {
                            query.limit = limit.toString();
                        }
                        if (orderDirection) {
                            query.direction = orderDirection;
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._client.apiCall({
                                url: "channels/" + channelId + "/subscriptions",
                                query: query,
                                scope: 'channel_subscriptions'
                            })];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, data.subscriptions.map(function (sub) { return new ChannelSubscription_1.default(sub, _this._client); })];
                    case 3:
                        e_1 = _a.sent();
                        if (e_1 instanceof errors_1.StatusCodeError && e_1.statusCode === 422) {
                            throw new NoSubscriptionProgram_1.default(channelId);
                        }
                        throw e_1;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    ChannelAPI.prototype.getChannelSubscriptionByUser = function (channel, byUser) {
        return __awaiter(this, void 0, void 0, function () {
            var channelId, userId, _a, e_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        channelId = UserTools_1.default.getUserId(channel);
                        userId = UserTools_1.default.getUserId(byUser);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        _a = ChannelSubscription_1.default.bind;
                        return [4 /*yield*/, this._client.apiCall({
                                url: "channels/" + channelId + "/subscriptions/" + userId,
                                scope: 'channel_check_subscription'
                            })];
                    case 2: return [2 /*return*/, new (_a.apply(ChannelSubscription_1.default, [void 0, _b.sent(),
                            this._client]))()];
                    case 3:
                        e_2 = _b.sent();
                        if (e_2 instanceof errors_1.StatusCodeError) {
                            if (e_2.statusCode === 404) {
                                throw new NotSubscribed_1.default(channelId, userId);
                            }
                            else if (e_2.statusCode === 422) {
                                throw new NoSubscriptionProgram_1.default(channelId);
                            }
                        }
                        throw e_2;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    ChannelAPI.prototype.startChannelCommercial = function (channel, length) {
        return __awaiter(this, void 0, void 0, function () {
            var channelId;
            return __generator(this, function (_a) {
                channelId = UserTools_1.default.getUserId(channel);
                return [2 /*return*/, this._client.apiCall({
                        url: "channels/" + channelId + "/commercial",
                        method: 'POST',
                        jsonBody: { length: length },
                        scope: 'channel_commercial'
                    })];
            });
        });
    };
    ChannelAPI.prototype.resetChannelStreamKey = function (channel) {
        return __awaiter(this, void 0, void 0, function () {
            var channelId;
            return __generator(this, function (_a) {
                channelId = UserTools_1.default.getUserId(channel);
                return [2 /*return*/, this._client.apiCall({
                        url: "channels/" + channelId + "/stream_key",
                        method: 'DELETE',
                        scope: 'channel_stream'
                    })];
            });
        });
    };
    __decorate([
        Decorators_1.Cached(3600)
    ], ChannelAPI.prototype, "getMyChannel", null);
    __decorate([
        Decorators_1.Cached(3600)
    ], ChannelAPI.prototype, "getChannel", null);
    __decorate([
        Decorators_1.ClearsCache('getChannel', 1)
    ], ChannelAPI.prototype, "updateChannel", null);
    __decorate([
        Decorators_1.Cached(3600)
    ], ChannelAPI.prototype, "getChannelEditors", null);
    __decorate([
        Decorators_1.Cached(30)
    ], ChannelAPI.prototype, "getChannelFollowers", null);
    __decorate([
        Decorators_1.Cached(30)
    ], ChannelAPI.prototype, "getChannelSubscriptions", null);
    __decorate([
        Decorators_1.Cached(3600)
    ], ChannelAPI.prototype, "getChannelSubscriptionByUser", null);
    __decorate([
        Decorators_1.ClearsCache('getMyChannel')
    ], ChannelAPI.prototype, "resetChannelStreamKey", null);
    ChannelAPI = __decorate([
        Decorators_1.Cacheable
    ], ChannelAPI);
    return ChannelAPI;
}(BaseAPI_1.default));
exports.default = ChannelAPI;
